File: sistemas_operativos_III/2023/docker/docker-compose/php-apache-mysql/proyectoFinal/Persistencia.php
<?php

class Persistencia {
  
  //El objeto $con va a manejar la conexión a BD
  private $con;
  private $DB_HOST = "localhost";
  private $DB_USER = "root";
  private $DB_PASS = "root";
  private $DB_NAME = "usuarios";
  private $CHAR_SET = "UTF8";
 
  //Constructor de la clase
  function __construct() {
    try {
      $this->con = $this->getConnection();
    }
    catch (PDOException $ex) {
      echo $ex->getMessage();
    }
  }

  //Conexón a la BD
  public function getConnection():PDO {
    $opt = [PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC]; 
    $pdo = new PDO("mysql:host={$this->DB_HOST};dbname={$this->DB_NAME};charset={$this->CHAR_SET}", $this->DB_USER, $this->DB_PASS, $opt);
    $pdo->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    return $pdo;
  }
  
  //Retorna los registros que están en la tabla usuario de la base de datos
  public function retornarRegistros():PDOStatement{
    $sql = "SELECT * FROM usuario;";
    $registrosBD = $this->con->query($sql);
    return $registrosBD;
  }

  //Permite dar de alta un trabajador en el Sistema
  public function altaUsuario(Trabajador $trabajador):bool {
    $bandera = false;
    try {
      $this->con->beginTransaction();
      $pst = $this->con->prepare("INSERT INTO usuario VALUES(:ci,:nombre,:apellido,:direccion,:email,:celular)");
      $pst->bindParam(':ci', $trabajador->getCi(), PDO::PARAM_STR);
      $pst->bindParam(':nombre', $trabajador->getNombre(), PDO::PARAM_STR);
      $pst->bindParam(':apellido', $trabajador->getApellido(), PDO::PARAM_STR);
      $pst->bindParam(':direccion', $trabajador->getDireccion(), PDO::PARAM_STR);
      $pst->bindParam(':email', $trabajador->getEmail(), PDO::PARAM_STR);
      $pst->bindParam(':celular', $trabajador->getCelular(), PDO::PARAM_STR);
      $pst->execute();
      $this->con->commit();
      $bandera = true; 
    } catch (PDOException $ex) {
      $this->con->rollback();
      echo $ex->getMessage();
      return $bandera;
    }
    return $bandera;
  }

  //Retorna si existe un trabajador en la base de datos
  public function existeTrabajador(String $ci):bool {
    $existe = false; 
    $sql = "SELECT * FROM usuario where ci = '{$ci}';";
    $registrosBD = $this->con->query($sql);
    $registrosBD=$registrosBD->fetch();
    if ($registrosBD){
      $existe = true;
    }
    return $existe;   
  }

  //Retorna un trabajador de la base de datos
  public function retornarRegistro(String $ci):PDOStatement {
    $sql = "SELECT * FROM usuario where ci = '{$ci}';";
    $registrosBD = $this->con->query($sql);
    return $registrosBD;
  }

  //Borrar un trabajador de la base de datos
  public function borrarRegistro(String $ci):bool {
    $val = $this->existeTrabajador($ci);
    $borrado = false;
    if($val){
      try {
        $this->con->beginTransaction();
        $pst = $this->con->prepare("DELETE FROM usuario WHERE ci = :ci");
        $pst->bindParam(":ci", $ci, PDO::PARAM_STR);
        $pst->execute();
        $this->con->commit();
        $borrado = true;
      } catch (PDOException $ex) {
        $this->con->rollback();
        echo $ex->getMessage();
        return $borrado; 
      }
    }
    return $borrado; 
  }

  public function modificarRegistro(Trabajador $trabajador):bool {
    $bandera = false;
    try {
      $this->con->beginTransaction();
      $pst = $this->con->prepare("UPDATE usuario SET nombre=:nombre,apellido=:apellido,direccion=:direccion,email=:email,celular=:celular WHERE ci=:ci");
      $pst->bindParam(':ci', $trabajador->getCi(), PDO::PARAM_STR);
      $pst->bindParam(':nombre', $trabajador->getNombre(), PDO::PARAM_STR);
      $pst->bindParam(':apellido', $trabajador->getApellido(), PDO::PARAM_STR);
      $pst->bindParam(':direccion', $trabajador->getDireccion(), PDO::PARAM_STR);
      $pst->bindParam(':email', $trabajador->getEmail(), PDO::PARAM_STR);
      $pst->bindParam(':celular', $trabajador->getCelular(), PDO::PARAM_STR);
      $pst->execute();
      $this->con->commit();
      $bandera = true; 
    } catch (PDOException $ex) {
      $this->con->rollback();
      echo $ex->getMessage();
      return $bandera;
    }
    return $bandera;
  }
//Fin de la clase  
}
?>

File: sistemas_operativos_III/2023/docker/docker-compose/php-apache-mysql/proyectoFinal/Persistencia.go
package persistencia

import (
	"database/sql"
	"errors"
	_ "github.com/go-sql-driver/mysql"
)

type Trabajador struct {
	Ci        string
	Nombre    string
	Apellido  string
	Direccion string
	Email     string
	Celular   string
}

type Persistencia struct {
	db *sql.DB
}

func NewPersistencia() (*Persistencia, error) {
	db, err := sql.Open("mysql", "root:root@tcp(localhost)/usuarios")
	if err != nil {
		return nil, err
	}

	return &Persistencia{db: db}, nil
}

func (p *Persistencia) Close() {
	p.db.Close()
}

func (p *Persistencia) RetornarRegistros() ([]Trabajador, error) {
	rows, err := p.db.Query("SELECT * FROM usuario")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var trabajadores []Trabajador

	for rows.Next() {
		var t Trabajador
		err = rows.Scan(&t.Ci, &t.Nombre, &t.Apellido, &t.Direccion, &t.Email, &t.Celular)
		if err != nil {
			return nil, err
		}
		trabajadores = append(trabajadores, t)
	}

	return trabajadores, nil
}

func (p *Persistencia) AltaUsuario(t Trabajador) error {
	_, err := p.db.Exec("INSERT INTO usuario VALUES (?, ?, ?, ?, ?, ?)", t.Ci, t.Nombre, t.Apellido, t.Direccion, t.Email, t.Celular)
	if err != nil {
		return err
	}
	return nil
}

func (p *Persistencia) ExisteTrabajador(ci string) (bool, error) {
	var count int
	err := p.db.QueryRow("SELECT COUNT(*) FROM usuario WHERE ci = ?", ci).Scan(&count)
	if err != nil {
		return false, err
	}
	return count > 0, nil
}

func (p *Persistencia) RetornarRegistro(ci string) (Trabajador, error) {
	var t Trabajador
	err := p.db.QueryRow("SELECT * FROM usuario WHERE ci = ?", ci).Scan(&t.Ci, &t.Nombre, &t.Apellido, &t.Direccion, &t.Email, &t.Celular)
	if err != nil {
		return Trabajador{}, err
	}
	return t, nil
}

func (p *Persistencia) BorrarRegistro(ci string) error {
	existe, err := p.ExisteTrabajador(ci)
	if err != nil {
		return err
	}
	if existe {
		_, err = p.db.Exec("DELETE FROM usuario WHERE ci = ?", ci)
		if err != nil {
			return err
		}
	}
	return nil
}

func (p *Persistencia) ModificarRegistro(t Trabajador) error {
	_, err := p.db.Exec("UPDATE usuario SET nombre = ?, apellido = ?, direccion = ?, email = ?, celular = ? WHERE ci = ?", t.Nombre, t.Apellido, t.Direccion, t.Email, t.Celular, t.Ci)
	if err != nil {
		return err
	}
	return nil
}
